csharp/sources/Valkey.Glide.InterOp/NativeClient.cs

```diff
@@ -0,0 +1,509 @@
+    {
+        if (buffer is null)
+            return;
+        nint ptr = (nint)(buffer - 1);
+        if (Marshal.ReadByte(ptr) == 0) // Is allocated on heap
+            Marshal.FreeCoTaskMem(ptr);
+    }
+
+    private static unsafe byte* MarshalBytes(byte[] bytes, byte* buffer = null, int bufferLength = 0)
+    {
+        if (buffer is null && bufferLength is not 0)
+            throw new ArgumentException("Buffer is null and bufferLength is not 0", nameof(buffer));
+        if (bufferLength is 0 || bytes.Length > bufferLength - 2)
+        {
+            nint ptr = Marshal.AllocHGlobal(bytes.Length + 2);
+            fixed (byte* bytesPtr = bytes)
+            {
+                Marshal.WriteByte(ptr, 0); // Is allocated on heap
+                for (int i = 0; i < bytes.Length; i++)
+                    Marshal.WriteByte(ptr + i + 1, bytesPtr[i]);
+                Marshal.WriteByte(ptr + bytes.Length + 1, 0);
+            }
+
+            return (byte*)ptr + 1;
+        }
+        else
+        {
+            nint ptr = (nint)buffer;
+            fixed (byte* bytesPtr = bytes)
+            {
+                Marshal.WriteByte(ptr, 1); // Is allocated on buffer (potentially stack)
+                for (int i = 0; i < bytes.Length; i++)
+                    Marshal.WriteByte(ptr + i + 1, bytesPtr[i]);
+                Marshal.WriteByte(ptr + bytes.Length + 1, 0);
+            }
+
+            return buffer + 1;
+        }
+    }
+
+    private static unsafe byte* MarshalUtf8String(string s, byte* buffer = null, int bufferLength = 0)
+    {
+        if (buffer is null && bufferLength is not 0)
+            throw new ArgumentException("Buffer is null and bufferLength is not 0", nameof(buffer));
+        byte[] utf8 = Encoding.UTF8.GetBytes(s);
+        if (bufferLength is 0 || utf8.Length > bufferLength - 2)
+        {
+            nint ptr = Marshal.AllocHGlobal(utf8.Length + 2);
+            fixed (byte* utf8Ptr = utf8)
+            {
+                Marshal.WriteByte(ptr, 0); // Is allocated on heap
+                for (int i = 0; i < utf8.Length; i++)
+                    Marshal.WriteByte(ptr + i + 1, utf8Ptr[i]);
+                Marshal.WriteByte(ptr + utf8.Length + 1, 0);
+            }
+
+            return (byte*)ptr + 1;
+        }
+        else
+        {
+            nint ptr = (nint)buffer;
+            fixed (byte* utf8Ptr = utf8)
+            {
+                Marshal.WriteByte(ptr, 1); // Is allocated on buffer (potentially stack)
+                for (int i = 0; i < utf8.Length; i++)
+                    Marshal.WriteByte(ptr + i + 1, utf8Ptr[i]);
+                Marshal.WriteByte(ptr + utf8.Length + 1, 0);
+            }
+
+            return buffer + 1;
+        }
+    }
+
+
+    private delegate void CommandCallbackDelegate([In] nint data, [In] int success, [In] Native.Value payload);
+
+    private static void CommandCallback([In] nint data, [In] int success, [In] Native.Value payload)
+    {
+        try
```

Comment:
Callback should be as fast as possible and have no exceptions - because it runs in rust's runtime. See how it works on `main` now.

Reply:
This callback performed not in dotnet runtime, but in tokio runtime. That runtime has only 1 thread so far and we should use that as least as possible. Throwing an exception there will cause a panic in rust part and termination of that thread. Existing running tasks won't complete and client may lose user's data.
In a rust->C# callback we have to copy data and all further handling of it should be done in async manner.